library/java/org/cronvoy/BidirectionalStreamBuilderImpl.java

```diff
@@ -0,0 +1,145 @@
+  // Priority of the stream. Default is medium.
+  @StreamPriority private int priority = STREAM_PRIORITY_MEDIUM;
+
+  private boolean delayRequestHeadersUntilFirstFlush;
+
+  // Request reporting annotations.
+  private Collection<Object> requestAnnotations;
+
+  private boolean trafficStatsTagSet;
+  private int trafficStatsTag;
+  private boolean trafficStatsUidSet;
+  private int trafficStatsUid;
+
+  /**
+   * Creates a builder for {@link BidirectionalStream} objects. All callbacks for generated {@code
+   * BidirectionalStream} objects will be invoked on {@code executor}. {@code executor} must not run
+   * tasks on the current thread, otherwise the networking operations may block and exceptions may
+   * be thrown at shutdown time.
+   *
+   * @param url the URL for the generated stream
+   * @param callback the {@link BidirectionalStream.Callback} object that gets invoked upon
+   *     different events occuring
+   * @param executor the {@link Executor} on which {@code callback} methods will be invoked
+   * @param cronetEngine the {@link CronetEngine} used to create the stream
+   */
+  BidirectionalStreamBuilderImpl(String url, BidirectionalStream.Callback callback,
+                                 Executor executor, CronvoyEngineBase cronetEngine) {
+    super();
+    if (url == null) {
+      throw new NullPointerException("URL is required.");
+    }
+    if (callback == null) {
+      throw new NullPointerException("Callback is required.");
+    }
+    if (executor == null) {
+      throw new NullPointerException("Executor is required.");
+    }
+    if (cronetEngine == null) {
+      throw new NullPointerException("CronetEngine is required.");
+    }
+    this.url = url;
+    this.callback = callback;
+    this.executor = executor;
+    this.cronetEngine = cronetEngine;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl setHttpMethod(String method) {
+    if (method == null) {
+      throw new NullPointerException("Method is required.");
+    }
+    httpMethod = method;
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl addHeader(String header, String value) {
+    if (header == null) {
+      throw new NullPointerException("Invalid header name.");
+    }
+    if (value == null) {
+      throw new NullPointerException("Invalid header value.");
+    }
+    requestHeaders.add(new AbstractMap.SimpleImmutableEntry<>(header, value));
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl setPriority(@StreamPriority int priority) {
+    this.priority = priority;
+    return this;
+  }
+
+  @Override
+  public BidirectionalStreamBuilderImpl
+  delayRequestHeadersUntilFirstFlush(boolean delayRequestHeadersUntilFirstFlush) {
+    this.delayRequestHeadersUntilFirstFlush = delayRequestHeadersUntilFirstFlush;
+    return this;
+  }
+
+  @Override
+  public ExperimentalBidirectionalStream.Builder addRequestAnnotation(Object annotation) {
+    if (annotation == null) {
```

Comment:
I've been pondering if we should add nullability lint checking and annotations to the java layer. If this is something that you think would be helpful to have, I can work on getting that set up for us

Reply:
We tried it in my previous teams. At the end, we concluded that it was too much of a hassle: to avoid lint errors, especially with generics, we saw SWEs doing some nasty hacking - did not payoff.